pandas/core/window/common.py

```diff
@@ -83,8 +83,22 @@ def dataframe_from_int_dict(data, frame_template):
                         # mypy needs to know columns is a MultiIndex, Index doesn't
                         # have levels attribute
                         arg2.columns = cast(MultiIndex, arg2.columns)
-                        result.index = MultiIndex.from_product(
-                            arg2.columns.levels + [result_index]
+                        # GH 21157: Equivalent to MultiIndex.from_product(
+                        # <unique combinations of arg2.columns.levels>, [result_index]
+                        # )
+                        # A normal MultiIndex.from_product will produce too many
+                        # combinations.
+                        idx_codes, idx_uniques = result_index.factorize()
```

Comment:
why are you not using `.from_arrays` here? this is really hard to understand *why* you are doing this at all.

Reply:
Good point. Simplified to use `from_arrays`.